Added new parameter to handle size of log file

GlaciErrDev opened the pull request:
- Added log_file_size option
  - Added functionality to truncate old logs in file

Commit:
Added new parameter to handle size of log file

  - Added log_file_size option
  - Added functionality to truncate old logs in file

joker-at-work commented on vspc/server.py:
```diff
@@ -195,6 +198,24 @@ async def save_to_log(self, uuid, data):
+            await read_obj.seek(start_point)
+            return await read_obj.read()
+
+    async def rewrite_file(self, path, data):
+        tmp_path = "tmp_" + path
+        async with aiofiles.open(tmp_path, "wb") as f:
+            await f.write(data)
+        await aio_os.remove(path)
+        await aio_os.rename(tmp_path, path)
```
This doesn't look threadsafe to me. Is it possible for another thread to write to the same file you're just replacing? Then we would lose data.

Commit:
Use only 'os.rename()' for atomic operation on file

joker-at-work commented:
> Use only for atomic operation on file

use only what?

GlaciErrDev commented:
Use only `rename`... it's disappeared...

joker-at-work commented:
Please add it again.

joker-at-work commented on vspc/server.py:
```diff
@@ -195,6 +198,24 @@ async def save_to_log(self, uuid, data):
         fpath = os.path.join(CONF.serial_log_dir, uuid)
         async with aiofiles.open(fpath, 'ab') as f:
             await f.write(data)
+        if CONF.log_file_size and os.path.getsize(fpath) > CONF.log_file_size:
```
Who awaits the `getsize()`?

GlaciErrDev replied:
everyone, it's a cheap operation, but better to use `ospath` from `aiofiles`

GlaciErrDev replied:
but I won't use it because it's not implemented in the library yet.

joker-at-work commented on vspc/server.py:
```diff
@@ -195,6 +198,24 @@ async def save_to_log(self, uuid, data):
+            truncated_data = await self.truncate_log_file(fpath, CONF.log_file_size)
+            await self.rewrite_file(fpath, truncated_data)
+
+    async def truncate_log_file(self, file_name, size):
+        async with aiofiles.open(file_name, "rb") as read_obj:
+            await read_obj.seek(0, os.SEEK_END)
+            pointer_location = await read_obj.tell()
+            start_point = pointer_location - size
+            await read_obj.seek(start_point)
+            return await read_obj.read()
```
The logs are line based, so I'd prefer not having half lines in there because of seeking. What do you think? Would it make sense to go further until the next line? (Maybe it's as easy as creating a file from it and doing a `readline()` before the final read.)

GlaciErrDev replied:
it was the first solution, but when you said that we don't care about past old logs I've decided just to seek to point of bytes. :) And yes, it's better when we have a whole line rather than a broken one...

joker-at-work commented on vspc/server.py:
```diff
@@ -195,6 +198,24 @@ async def save_to_log(self, uuid, data):
+            truncated_data = await self.truncate_log_file(fpath, CONF.log_file_size)
+            await self.rewrite_file(fpath, truncated_data)
+
+    async def truncate_log_file(self, file_name, size):
```
I find the name confusing. It doesn't actually truncate the log file, but just returns the truncated data. From the name it sounded like it would also change the file. Can we find a better name or move the `rewrite_file()` into the function so it actually changes the file? I don't see it used anywhere else.

GlaciErrDev replied:
Sure, sounds like it is better to move `rewrite_file` to `truncate_log_file`

joker-at-work commented on vspc/server.py:
```diff
@@ -214,7 +214,6 @@ async def rewrite_file(self, path, data):
         async with aiofiles.open(tmp_path, "wb") as f:
             await f.write(data)
-        await aio_os.remove(path)
         await aio_os.rename(tmp_path, path)
```
At least for Python's `os`, `rename()` doesn't overwrite. Instead, the [docs](https://docs.python.org/3/library/os.html#os.rename) talk about using `replace()`. Could you link the aiofile docs here/in the commit for future reference?

GlaciErrDev replied:
from docs link `rename` -> `If both are files, dst it will be replaced silently if the user has permission.`
`replace` is doing the same but it's a cross-platform operation -> `If you want cross-platform overwriting of the destination, use replace().`

joker-at-work replied:
You're right. I only read until "will raise OSError" and missed the "in a number of cases". :+1:

joker-at-work reviewed (commented):
Please rebase onto latest master to get rid of the additional commits.

Commit:
Added fixes regarding comments

  - Moved 'rewrite_file' to 'truncate_log_file'
  - After `seek` to a needed point we are reading line to prevent
    saving broken line to a new file

GlaciErrDev commented on etc/vspc.conf.sample:
```diff
@@ -6,4 +6,4 @@ cert = cert.pem
 uri = vmware-vspc
 serial_log_dir = /opt/vmware/vspc
-
+log_file_size = 1073741824
```
I'm just wondering, are you agree to use 1GB for the threshold to start to resize the log file?

joker-at-work replied:
This is just an example, right? I wouldn't save more than 50MB per VM in production. Even that seems quite big to me.

GlaciErrDev replied:
yes
